### PR TITLE
ksync 0.4.5 (new formula)

### PR DESCRIPTION
**Pull request by timfallmk**
New formula for Ksync, a tool for real-time sync of dev environments with a kubernetes cluster

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/ksync.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    :tag  => "0.4.5"
+  sha256 "bd47c5b5418bfc6d2f08abe855a52d4189f8bf380747cdd4e5c1a3ca0a87498a"
+  license "Apache-2.0"
+
+  depends_on "go" => :build
+
+  def install
+    # Set some stamps and build local CLI binary
+    time = `date +%Y%m%dT%H%M%S`.chomp
+    goversion = `go version | awk '{ print $3 }'`
```
Comment (alebcay) [inline Formula/ksync.rb:14]: I think something like this would probably be more preferable:
```
goversion = "go#{Formula["go"].version.to_s}"

Comment (timfallmk) [reply Formula/ksync.rb:14]: Readability?

Comment (alebcay) [reply Formula/ksync.rb:14]: That's a fair point. I'll leave the call to a different maintainer, although I'll put the option out here.

Comment (timfallmk) [reply Formula/ksync.rb:14]: Happy to do it, just want to match convention and be consistent with the binary that is distributed via non-homebrew methods.

File: Formula/ksync.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+class Ksync < Formula
+  desc "Sync files between your local system and a kubernetes cluster"
+  homepage "https://ksync.github.io/ksync/"
+  url "https://github.com/ksync/ksync.git",
+    :tag  => "0.4.5"
+  sha256 "bd47c5b5418bfc6d2f08abe855a52d4189f8bf380747cdd4e5c1a3ca0a87498a"
+  license "Apache-2.0"
+
```
Comment (timfallmk) [inline Formula/ksync.rb:8]: I get an audit error about this whitespace, but I see it in many other formulae? Is this incorrect?

File: Formula/ksync.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  depends_on "go" => :build
+
+  def install
+    # Set some stamps and build local CLI binary
+    time = `date +%Y%m%dT%H%M%S`.chomp
+    goversion = `go version | awk '{ print $3 }'`
+    commit = Utils.safe_popen_read("git", "rev-parse", "--short", "HEAD").chomp
+    system "go", "build", "-ldflags",
+              "-w -X github.com/ksync/ksync/pkg/ksync.GitCommit=#{commit} -X github.com/ksync/ksync/pkg/ksync.GitTag=#{version} -X github.com/ksync/ksync/pkg/ksync.BuildDate=#{time} -X github.com/ksync/ksync/pkg/ksync.VersionString=Homebrew -X github.com/ksync/ksync/pkg/ksync.GoVersion=#{goversion}",
```
Comment (timfallmk) [inline Formula/ksync.rb:17]: Audit error about the length of this line, but it needs to be passed as a _single_ argument to `-ldflag` to work. Suggestions?

File: Formula/ksync.rb (hunk 1 above)
Comment (SMillerDev) [reply Formula/ksync.rb:14]: I think the suggested improvement is more in line with homebrew state of the art.

File: Formula/ksync.rb (hunk 3 above)
Comment (SMillerDev) [reply Formula/ksync.rb:17]: Everywhere you have a space you can also do `",\n"` that should fix it.

File: Formula/ksync.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+              *std_go_args,
+              "github.com/ksync/ksync/cmd/ksync"
+  end
+
+  test do
+    # Basic build test. Potential for more sophisticated tests in the future
+    # Initialize the local client and see if it completes successfully
+    system "ksync", "init", "--local", "--skip-checks"
```
Comment (SMillerDev) [inline Formula/ksync.rb:25]: If this creates files, please assert that they exist after. If it has a text output, please use `shell_output` and `assert_match` to see what the output is.

File: Formula/ksync.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+class Ksync < Formula
+  desc "Sync files between your local system and a kubernetes cluster"
+  homepage "https://ksync.github.io/ksync/"
+  url "https://github.com/ksync/ksync.git",
+    :tag  => "0.4.5"
```
Comment (SMillerDev) [inline Formula/ksync.rb:5]: This needs to have a revision as well, to detect retags.

File: Formula/ksync.rb (hunk 3 above)
Comment (alebcay) [reply Formula/ksync.rb:17]: You could also try something like:
```
"-w -X github.com/ksync/ksync/pkg/ksync.GitCommit=#{commit}" \
"-X github.com/ksync/ksync/pkg/ksync.GitTag=#{version}" \
"-X github.com/ksync/ksync/pkg/ksync.BuildDate=#{time}" \
"-X github.com/ksync/ksync/pkg/ksync.VersionString=Homebrew" \
"-X github.com/ksync/ksync/pkg/ksync.GoVersion=#{goversion}"
```

The `\` operator will concatenate the multiple lines into a single argument. You could also try something along the lines of:
```
ldflags = %W[
  -w -X github.com/ksync/ksync/pkg/ksync.GitCommit=#{commit}
  -X github.com/ksync/ksync/pkg/ksync.GitTag=#{version}
  -X github.com/ksync/ksync/pkg/ksync.BuildDate=#{time}
  -X github.com/ksync/ksync/pkg/ksync.VersionString=Homebrew
  -X github.com/ksync/ksync/pkg/ksync.GoVersion=#{goversion}
]

[...]

system "go", "build", "-ldflags", ldflags.join(" ")
```

File: Formula/ksync.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+  # PLACEHOLDER: Need to find a way of running cleanup before uninstall
+  # def uninstall
+  #   system "ksync", "clean", "--nuke"
+  # end
+
+  test do
+    # Basic build test. Potential for more sophisticated tests in the future
+    # Initialize the local client and see if it completes successfully
+    system "ksync", "--help"
```
Comment (SMillerDev) [inline Formula/ksync.rb:34]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

File: Formula/ksync.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+              "-w -X github.com/ksync/ksync/pkg/ksync.GitCommit=#{commit} \
+              -X github.com/ksync/ksync/pkg/ksync.GitTag=#{version} \
+              -X github.com/ksync/ksync/pkg/ksync.BuildDate=#{time} \
+              -X github.com/ksync/ksync/pkg/ksync.VersionString=Homebrew \
+              -X github.com/ksync/ksync/pkg/ksync.GoVersion=#{goversion}",
+              *std_go_args,
+              "github.com/ksync/ksync/cmd/ksync"
+  end
+
+  # PLACEHOLDER: Need to find a way of running cleanup before uninstall
```
Comment (SMillerDev) [inline Formula/ksync.rb:26]: Why is that needed?

File: Formula/ksync.rb (hunk 6 above)
Comment (timfallmk) [reply Formula/ksync.rb:34]: The problem here is that all commands (including `version`) check for a valid cluster context and connection first. Unless it passes basic connection tests, the command won't work. The only exception is `help` for top level commands.

In our CI tests, we use a local container cluster (using `k3d`) to achieve this, so there's been no need to develop a completely local test. Even the `ksync init --local` command will check for basic health of a connection before completing.

I'm open to other suggestions! (cc @SMillerDev)

File: Formula/ksync.rb (hunk 7 above)
Comment (timfallmk) [reply Formula/ksync.rb:26]: Self reminder. I can find no formula syntax for pre-uninstall cleanup steps.

File: Formula/ksync.rb (hunk 6 above)
Comment (SMillerDev) [reply Formula/ksync.rb:34]: > If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
Same if the software requires a virtual machine, docker instance, etc. to be running.

File: Formula/ksync.rb (hunk 7 above)
Comment (SMillerDev) [reply Formula/ksync.rb:26]: That's because there isn't any. What would be the result if this isn't ran?

Comment (timfallmk) [reply Formula/ksync.rb:26]: Two things of importance:
- Local config and binaries are not cleaned up
- Remote components (cluster side) are not removed

File: Formula/ksync.rb (hunk 6 above)
Comment (timfallmk) [reply Formula/ksync.rb:34]: You would like to check for the _invalidity_ of a cluster connection? I'm not sure what that would accomplish other than checking the binary is intact (same as `help`) but it can easily be implemented.

File: Formula/ksync.rb (hunk 7 above)
Comment (SMillerDev) [reply Formula/ksync.rb:26]: The local files will be removed by brew because we're managing the package.

As for the remote files, I think you should just include that step in your documentation.

File: Formula/ksync.rb (hunk 6 above)
Comment (SMillerDev) [reply Formula/ksync.rb:34]: Yes, checking if the packaging is successful is what we want to test. And more often than not checking `--help` doesn't provide us with that information.

File: Formula/ksync.rb (hunk 7 above)
Comment (timfallmk) [reply Formula/ksync.rb:26]: Would brew remove files it doesn't know about? (the `init` command creates said files)

Comment (timfallmk) [reply Formula/ksync.rb:26]: Maybe adding a warning to the install would be good enough for the moment

File: Formula/ksync.rb (hunk 6 above)
Comment (timfallmk) [reply Formula/ksync.rb:34]: In this case I might suggest that it's the same. The only step of relevance here is the build step during the install (if it's not pulling from the bottle, as it should normally be). Since it's a single binary, and command is parsed through it, and would indicate that the build has been successful. Depending on the condition of the environment when `brew test` is run, the expected output could vary.

File: Formula/ksync.rb (hunk 7 above)
Comment (SMillerDev) [reply Formula/ksync.rb:26]: Brew would not remove files it doesn't know about. But it's up to users to learn how the tool works. Unless there's something that the user needs to know only for homebrew you shouldn't print a warning though.

It sounds like you want to clean up a staging area, and that's up to users.

File: Formula/ksync.rb (hunk 6 above)
Comment (gromgit) [reply Formula/ksync.rb:34]: > In this case I might suggest that it's the same.

It's not the same at all. Printing help and version texts are incidental functions, generally implemented as short-circuits that go nowhere near the main code path. A binary that tells you what version it is, but crashes when you try to use it for its intended purpose, is useless. That's why `brew test` needs to exercise some portion of the main functionality, even if it's known to fail.

In software development terms, if the only unit tests you ever wrote were for `help` and `version`, who would trust your code?

Comment (timfallmk) [reply Formula/ksync.rb:34]: I understand your point, but what I'm saying is a little different. All functions (and `go test` tests btw) require an active cluster context and healthy connection to execute. The check for these is execute _before_ any of the underlying command is called, meaning that any command (with the exception of `help`) _including_ `version` will first check for this, which as you said, goes nowhere near the "main" code. If the test is to check the validity of the compilation, then running `ksync help` is identical to say running `ksync init` or anything else.

However, since this seems to be a point of contention, I've changed it in the next commit. Note that this hits the same code as any function would (with the exception of help, which actually does more) and stops.

File: Formula/ksync.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+class Ksync < Formula
+  desc "Sync files between your local system and a kubernetes cluster"
+  homepage "https://ksync.github.io/ksync/"
+  url "https://github.com/ksync/ksync.git",
+    :tag      => "0.4.5",
+    :revision => "9f40bf134329814a57e1a58d73b84761a2b06c73"
+  sha256 "bd47c5b5418bfc6d2f08abe855a52d4189f8bf380747cdd4e5c1a3ca0a87498a"
+  license "Apache-2.0"
+  depends_on "go" => :build
```
Comment (SMillerDev) [inline Formula/ksync.rb:9]:
```suggestion
  license "Apache-2.0"

  depends_on "go" => :build
```

File: Formula/ksync.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+  def install
+    # Set some stamps and build local CLI binary
```
Comment (SMillerDev) [inline Formula/ksync.rb:12]: No need for the comment, the code is clear

File: Formula/ksync.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+              -X github.com/ksync/ksync/pkg/ksync.BuildDate=#{time} \
+              -X github.com/ksync/ksync/pkg/ksync.VersionString=Homebrew \
+              -X github.com/ksync/ksync/pkg/ksync.GoVersion=#{goversion}",
+              *std_go_args,
+              "github.com/ksync/ksync/cmd/ksync"
+  end
+
+  test do
+    expected = 'level=fatal msg="Unable to get config for context (). Does \'kubectl --context= cluster-info\' work?"'
+    assert_equal "#{expected}", shell_output("#{bin}/ksync init --local --log-level debug | awk '{print substr($0, index($0,$2))}'").split("\n").pop
```
Comment (SMillerDev) [inline Formula/ksync.rb:28]: This should be enough.
```suggestion
    expected = "level=fatal msg=\"Unable to get config for context (). Does 'kubectl --context= cluster-info' work?\""
    assert_equal "#{expected}", shell_output("#{bin}/ksync init --local --log-level debug")
```

File: Formula/ksync.rb (hunk 8 above)
Comment (timfallmk) [reply Formula/ksync.rb:9]: As I mentioned above, this gives me an audit error about unneeded whitespace

Comment (SMillerDev) [reply Formula/ksync.rb:9]: That might have been a temporary glitch. Try updating to the latest brew version and checking again.

File: Formula/ksync.rb (hunk 10 above)
Comment (timfallmk) [reply Formula/ksync.rb:28]: `awk` (or some method of string substitution) is needed to trim the timestamp off the front of the error message. Since the audit test seems to dislike `awk`, I'm looking at using native ruby methods.

Comment (SMillerDev) [reply Formula/ksync.rb:28]: Just use `assert_match` then. It'll take a regex (or just match a string anywhere)

**Comment by timfallmk** [comment]
This should be ready to go. Tests and audits passing, and all comments addressed.

File: Formula/ksync.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  url "https://github.com/ksync/ksync.git",
+    :tag      => "0.4.5",
+    :revision => "9f40bf134329814a57e1a58d73b84761a2b06c73"
+  sha256 "bd47c5b5418bfc6d2f08abe855a52d4189f8bf380747cdd4e5c1a3ca0a87498a"
+  license "Apache-2.0"
+
+  depends_on "go" => :build
+
+  def install
+    time = `date +%Y%m%dT%H%M%S`.chomp
```
Comment (SMillerDev) [inline Formula/ksync.rb:13]:
```suggestion
    time = Utils.safe_popen_read("date", "+%Y%m%dT%H%M%S").chomp
```

**Review by SMillerDev — COMMENTED** [review]
After this I think it's good to go. Please squash your commits into one with the same title as the PR.

**Review by SMillerDev — APPROVED** [review]
Thanks @timfallmk  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉
